### PR TITLE
Atomic Linked List

### PR DESCRIPTION
**Pull request by zrthxn**
Formatter and Tests

### CODE REVIEW

File: new.sh [hunk 1]
```diff
@@ -3,15 +3,22 @@ classname=$1
...
 touch __init__.py ${classname}_formatter.py test_${classname}.py
-touch ${classname}.cpp
+
+cat << EOF > ${classname}.cpp
+#include <iostream>
+
+int main() {
+  return 0;
+}
+EOF
```
Comment (a20012251) [inline new.sh:13]: this should be in a different PR

File: src/atomicll/atomicll.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+int main() {
+  AtomicLinkedList<char> ll;
+  ll.insertHead('A');
+  ll.insertHead('B');
+  ll.insertHead('C');
+
+  AtomicLinkedList<IPAddress> li;
+  li.insertHead(IPAddress("10.0.0.100"));
+  li.insertHead(IPAddress("10.0.1.200"));
```
Comment (a20012251) [inline src/atomicll/atomicll.cpp:15]: nice!!!

File: src/atomicll/atomicll_formatter.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+class AtomicLLSyntheticFormatter:
+    count = None
+    __end_of_list_address = 0
+
+    def __init__(self, valobj, dict):
+        self.valobj = valobj
+        # ## We dont get a "has_prev" value in __init__ and we're not sure if its
+        # ## because the AtomicLL is always singly linked 
+        self.has_prev = False
```
Comment (a20012251) [inline src/atomicll/atomicll_formatter.py:12]: remove this

File: src/atomicll/atomicll_formatter.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
+import os
+import lldb
+
+class AtomicLLSyntheticFormatter:
+    count = None
+    __end_of_list_address = 0
```
Comment (a20012251) [inline src/atomicll/atomicll_formatter.py:6]: remove this

File: src/atomicll/atomicll_formatter.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        self.has_prev = False
+    
+    def next_node(self, node):
+        return (node
+            .GetChildMemberWithName('hook')
+                .GetChildMemberWithName('next')
+        )
+
+    def is_valid(self, node):
+        return self.value(self.next_node(node)) != self.__end_of_list_address
```
Comment (a20012251) [inline src/atomicll/atomicll_formatter.py:21]: != 0

File: src/atomicll/atomicll_formatter.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            # ## doesn't get a "has_prev" value and we're not sure if its
+            # ## because the AtomicLL is always singly linked 
+            # if self.has_prev:
+            #     prev_val = self.prev.GetValueAsUnsigned(0)
+            #     if prev_val == 0:
+            #         return 0
+            #     if next_val == self.node_address:
+            #         return 0
+            #     if next_val == prev_val:
+            #         return 1
```
Comment (a20012251) [inline src/atomicll/atomicll_formatter.py:48]: delete

**Review by a20012251 — CHANGES_REQUESTED** [review]
just remove some stuff and then approved!

File: new.sh (hunk 1 above)
Comment (zrthxn) [reply new.sh:13]: Will do

**Comment by zrthxn** [comment]
@a20012251 Done the changes
Also new PR for `new.sh` -> https://github.com/a20012251/lldb_folly_formatter/pull/10
